File: python_src/src/helps_bots/mv_dots.py
"""
Move dots utilities for references
"""

import re
from typing import Optional


def move_dots_before_refs(text: str, lang: Optional[str] = None) -> str:
    """Move dots before references to after references

    Args:
        text: Text containing references and punctuation
        lang: Language code

    Returns:
        Text with dots moved
    """
    dot = r"\.,。।"

    if lang == "hy":
        dot = r"\.,。։।"

    # Pattern to match dots before refs
    regline = r"((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)"
    pattern = rf"([{dot}]+)\s*{regline}"
    replacement = r"\2\1"

    text = re.sub(pattern, replacement, text, flags=re.MULTILINE)

    return text


def move_dots_after_refs(text: str, lang: Optional[str] = None) -> str:
    """Move dots after references (keep punctuation after closing tags)

    Args:
        text: Text containing references and punctuation
        lang: Language code

    Returns:
        Text with dots positioned after refs
    """
    dot = r"\.,。।"

    if lang == "hy":
        dot = r"\.,。։:"

    regline = r"((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)"
    pattern = rf"({regline}+)([{dot}]+)"
    replacement = r"\2\1"

    text = re.sub(pattern, replacement, text, flags=re.MULTILINE)

    return text

File: fix_src/helps_bots/mv_dots.php
<?php

namespace WpRefs\MovesDots;

/*
usage:

use function WpRefs\MovesDots\move_dots_after_refs;

*/

function move_dots_before_refs(string $text, string $lang): string
{
    // Define punctuation marks based on language
    $punctuation = '\.,،';

    // Pattern to match references followed by punctuation
    // This pattern handles one or more ref tags followed by punctuation
    $pattern = '/((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)([\.,،])/su';
    $pattern = '/((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)([' . $punctuation . ']+)/su';

    // Replace by moving punctuation before the reference(s)
    $result = preg_replace_callback($pattern, function ($matches) {
        // Handle multiple dots by replacing with a single dot
        $punctuation = $matches[2];
        if (substr_count($punctuation, '.') > 1) {
            $punctuation = '.';
        }
        return $punctuation . ' ' . trim($matches[1]);
    }, $text);

    return $result;
}

function move_dots_after_refs($newtext, $lang)
{
    // ---
    // echo_test("move_dots_after_refs\n");
    // ---
    $dot = "\.,。।";
    // ---
    if ($lang === "hy") {
        $dot = "\.,。։।:";
    }
    // ---
    $regline = "((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)";
    // ---
    $pattern = "/([" . $dot . "]+)\s*" . $regline . "/mu";
    $replacement = "$2$1";
    // ---
    $newtext = preg_replace($pattern, $replacement, $newtext);
    // ---
    return $newtext;
}
